### PR TITLE
VM refactor: more OS stack usage reductions

### PR DESCRIPTION
**Pull request by chfast**
- VM instruction pointer is now uint64_t.
- CALL params are placed on stack instead of heap. That reduces OS stack usage by ~30%.

### CODE REVIEW

File: libevm/VM.cpp [hunk 1]
```diff
@@ -535,16 +531,16 @@ bytesConstRef VM::execImpl(u256& io_gas, ExtVMFace& _ext, OnOpFunc const& _onOp)
 			m_stack.pop_back();
 			break;
 		case Instruction::JUMP:
-			nextPC = m_stack.back();
-			if (find(m_jumpDests.begin(), m_jumpDests.end(), (uint64_t)nextPC) == m_jumpDests.end() || nextPC > numeric_limits<uint64_t>::max() )
+			nextPC = static_cast<uint64_t>(m_stack.back());
+			if (find(m_jumpDests.begin(), m_jumpDests.end(), nextPC) == m_jumpDests.end() || m_stack.back() > numeric_limits<uint64_t>::max())
```
Comment (chriseth) [inline libevm/VM.cpp]: I think you could use std::binary_search here.

File: libevm/VM.cpp [hunk 2]
```diff
@@ -598,7 +594,7 @@ bytesConstRef VM::execImpl(u256& io_gas, ExtVMFace& _ext, OnOpFunc const& _onOp)
 			break;
 		case Instruction::CREATE:
 		{
-			u256 endowment = m_stack.back();
+			auto& endowment = m_stack.back();
```
Comment (chriseth) [inline libevm/VM.cpp]: const ref?

File: libevm/VM.cpp [hunk 3]
```diff
@@ -614,16 +610,14 @@ bytesConstRef VM::execImpl(u256& io_gas, ExtVMFace& _ext, OnOpFunc const& _onOp)
 		case Instruction::CALL:
 		case Instruction::CALLCODE:
 		{
-			if (!callParams)
-				callParams.reset(new CallParameters);
-
-			callParams->gas = m_stack.back();
+			CallParameters callParams;
```
Comment (chriseth) [inline libevm/VM.cpp]: Interesting, I would have said this change takes more space on the stack...

Comment (chfast) [inline libevm/VM.cpp]: Me too. But it is exactly the opposite.

File: libevm/VM.cpp [hunk 4]
```diff
@@ -45,165 +45,161 @@ static array<InstructionMetric, 256> metrics()
...
 }
 
-bytesConstRef VM::execImpl(u256& io_gas, ExtVMFace& _ext, OnOpFunc const& _onOp)
+void VM::checkRequirements(u256& io_gas, ExtVMFace& _ext, OnOpFunc const& _onOp, Instruction _inst)
 {
-	// Reset leftovers from possible previous run
-	m_curPC = 0;
-	m_jumpDests.clear();
+	static const auto c_metrics = metrics();
+	auto& metric = c_metrics[static_cast<size_t>(_inst)];
```
Comment (chriseth) [inline libevm/VM.cpp]: Why not cast to `uint8_t`?

**Comment by chfast** [comment]
All issues fixed.

**Pull request merged**